include/aie/Dialect/AIEVec/Transforms/FoldMulAddChainToConvOp.h

```diff
@@ -0,0 +1,511 @@
+      if (dist > 4) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+
+      dists.push_back(dist);
+      if (curUpdOp.getSource() != nextUpdOp.getSource()) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+
+      MemRefType curMemRefType =
+          curUpdOp.getSource().getType().cast<MemRefType>();
+      MemRefType nextMemRefType =
+          nextUpdOp.getSource().getType().cast<MemRefType>();
+
+      ArrayRef<int64_t> curSizes = curMemRefType.getShape();
+      ArrayRef<int64_t> nextSizes = nextMemRefType.getShape();
+      if (curSizes.size() != nextSizes.size()) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+
+      AffineExpr curLinearAccess =
+          constructLinearizedAffineExprForUPDOp(curUpdOp);
+      AffineExpr nextLinearAccess =
+          constructLinearizedAffineExprForUPDOp(nextUpdOp);
+      if (!curLinearAccess || !nextLinearAccess) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+
+      AffineExpr curBase, nextBase;
+      int32_t curOffset, nextOffset;
+
+      // Get the base and offset from linear access expr
+      std::tie(curBase, curOffset) = extractBaseAndOffset(curLinearAccess);
+      std::tie(nextBase, nextOffset) = extractBaseAndOffset(nextLinearAccess);
+      if (curBase != nextBase) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+
+      dist = nextOffset - curOffset;
+      if (dist != 1) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+      dists.push_back(dist);
+
+      if ((xDist != -1 && xDist != dists[0]) ||
+          (zDist != -1 && zDist != dists[1])) {
+        if (fusedOps.size() < 2) {
+          return false;
+        }
+        refreshFusedGroups(defTuple, nextMulOp, fusedOps, groupFusedOps, curIdx,
+                           curUpdOp, curMulOp);
+        continue;
+      }
+
+      xDist = dists[0];
+      zDist = dists[1];
+      dupFactor = dists[0];
```

Comment:
For the conv2d case, it could be that for the first line we determing dupFactor=1, but the second line could set dupFactor=2. My point is that this code is overwriting the dupfactor computed in previous iterations. Is this leading to problems in the caller? Seems to me that we need to check that the dupfactor computed in sucessive iterations is the same.